<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="nfportalutils">
<title>Bringing Portal Data to Other Platforms: cBioPortal • nfportalutils</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Bringing Portal Data to Other Platforms: cBioPortal">
<meta property="og:description" content="nfportalutils">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">nfportalutils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9500</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/access-utilities.html">Access utilities</a>
    <a class="dropdown-item" href="../articles/annotate-nf-processed-data.html">Annotating nextflow processed data</a>
    <a class="dropdown-item" href="../articles/bringing-portal-data-to-other-platforms-cbioportal.html">Bringing Portal Data to Other Platforms: cBioPortal</a>
    <a class="dropdown-item" href="../articles/portal-tables-utils.html">Portal tables utils</a>
    <a class="dropdown-item" href="../articles/revalidation-workflows.html">Revalidation workflows</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nf-osi/nfportalutils/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Bringing Portal Data to Other Platforms: cBioPortal</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nf-osi/nfportalutils/blob/HEAD/vignettes/bringing-portal-data-to-other-platforms-cbioportal.Rmd" class="external-link"><code>vignettes/bringing-portal-data-to-other-platforms-cbioportal.Rmd</code></a></small>
      <div class="d-none name"><code>bringing-portal-data-to-other-platforms-cbioportal.Rmd</code></div>
    </div>

    
    
<p><strong>Document Status:</strong> Draft<br><strong>Estimated Reading Time:</strong> 8 min</p>
<div class="section level2">
<h2 id="special-acknowledgments">Special acknowledgments<a class="anchor" aria-label="anchor" href="#special-acknowledgments"></a>
</h2>
<p>Functionality demonstrated in this vignette benefited greatly from
code originally written by <a href="https://github.com/hhunterzinck" class="external-link">hhunterzinck</a>.</p>
</div>
<div class="section level2">
<h2 id="intro">Intro<a class="anchor" aria-label="anchor" href="#intro"></a>
</h2>
<p>This describes how to package some Synapse processed data as a
cBioPortal study dataset. A cBioPortal study contains one or more data
types, see <a href="https://docs.cbioportal.org/file-formats/" class="external-link">cBioPortal docs</a>.
The current API covers creating a cBioPortal study with a subset of data
types relevant to the NF workflow (so not all data types). The design
has been inspired by and should feel somewhat like working with the R
package <a href="https://github.com/r-lib/usethis" class="external-link">usethis</a>, and data
types can be added to the study package interactively.</p>
<p>Though there is some checking depending on the data type, final <a href="#Validation">validation</a> with the official cBioPortal
validation tools/scripts should still be run.</p>
<p>Breaking changes are possible as the API is still in development.</p>
</div>
<div class="section level2">
<h2 id="set-up">Set up<a class="anchor" aria-label="anchor" href="#set-up"></a>
</h2>
<p>First load the <code>nfportalutils</code> package and log in. The
recommended default usage of <code>syn_login</code> is to use it without
directly passing in credentials. Instead, have available the
<code>SYNAPSE_AUTH_TOKEN</code> environment variable with your token
stored therein.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/nf-osi/nfportalutils" class="external-link">nfportalutils</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/syn_login.html">syn_login</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-a-new-study-dataset">Create a new study dataset<a class="anchor" aria-label="anchor" href="#create-a-new-study-dataset"></a>
</h2>
<p>First create the study dataset “package” where we can put together
the data. Each study dataset combines multiple data types – clinical,
gene expression, gene variants, etc.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/cbp_new_study.html">cbp_new_study</a></span><span class="op">(</span>cancer_study_identifier <span class="op">=</span> <span class="st">"npst_nfosi_ntap_2022"</span>,</span>
<span>              name <span class="op">=</span> <span class="st">"Plexiform Neurofibroma and Neurofibroma (Pratilas 2022)"</span>,</span>
<span>              citation <span class="op">=</span> <span class="st">"TBD"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="add-data-types-to-study">Add data types to study<a class="anchor" aria-label="anchor" href="#add-data-types-to-study"></a>
</h2>
<p>Data types can be most easily added <em>in any order</em> using the
<code>cbp_add*</code> functions. These functions download data files and
create the meta for them.</p>
<p>Note that:</p>
<ul>
<li>These should be run with the working directory set to the study
dataset directory as set up above to ensure consistent metadata.</li>
<li>
<strong>Defaults are for known NF-OSI processed data
outputs</strong>.</li>
<li>If these defaults don’t apply because of changes in the scenario,
take a look at the lower-level utils <code>make_meta_*</code> or edit
the files manually after.</li>
<li>Data types can vary in how much additional work is needed in
remapping, reformatting, custom sanity checks, etc.</li>
</ul>
<div class="section level3">
<h3 id="add-mutations-data">Add mutations data<a class="anchor" aria-label="anchor" href="#add-mutations-data"></a>
</h3>
<ul>
<li>
<code>maf_data</code> references a final merged maf output file from
the NF-OSI processing pipeline OK for public release.</li>
<li>This data file type requires no further modifications except
renaming.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">maf_data</span> <span class="op">&lt;-</span> <span class="st">"syn36553188"</span></span>
<span></span>
<span><span class="fu">add_cbp_maf</span><span class="op">(</span><span class="va">maf_data</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-copy-number-alterations-cna-data">Add copy number alterations (CNA) data<a class="anchor" aria-label="anchor" href="#add-copy-number-alterations-cna-data"></a>
</h3>
<ul>
<li>
<code>cna_data</code> is expected to be a <code>.seg</code> file on
Synapse.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">cna_data</span> <span class="op">&lt;-</span> <span class="st">"syn********"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/cbp_add_cna.html">cbp_add_cna</a></span><span class="op">(</span><span class="va">cna_data</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-expression-data">Add expression data<a class="anchor" aria-label="anchor" href="#add-expression-data"></a>
</h3>
<ul>
<li>
<code>expression_data</code> is expected to be a <code>.txt</code>
called <code>gene_tpm.tsv</code> file on Synapse.</li>
<li>The NF-OSI default includes including the raw expression data as
well, called <code>gene_counts.tsv</code>, but this can be omitted.</li>
<li>These NF-OSI outputs will be somewhat modified in translation to
have the required headers.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mrna_data</span> <span class="op">&lt;-</span> <span class="st">"syn********"</span></span>
<span><span class="va">mrna_data_raw</span> <span class="op">&lt;-</span> <span class="st">"syn********"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/cbp_add_expression.html">cbp_add_expression</a></span><span class="op">(</span><span class="va">mrna_data</span>,</span>
<span>                   expression_data_raw <span class="op">=</span> <span class="va">mrna_data_raw</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-clinical-data">Add clinical data<a class="anchor" aria-label="anchor" href="#add-clinical-data"></a>
</h3>
<ul>
<li>
<code>clinical_data</code> is a prepared clinical data table already
subsetted to those released in this study, or pass in a query that can
be used for subsetting if using a full clinical database table. For
example, the full clinical cohort comprises patients 1-50, but this
study dataset consists of available and releasable data only for
patients 1-20 for expression data and data patients 15-20 for cna data.
Here, <code>clinical_data</code> can be a smaller table of just those
1-30, or it can be the original table but pass in a suitable additional
filter, e.g. <code>where release = 'batch1'</code>.</li>
<li>Clinical data requires mapping to be as consistent with other public
datasets as possible. <code>ref_map</code> defines the mapping of
clinical variables from the NF-OSI data dictionary to cBioPortal’s. Only
variables in the mapping are exported to cBioPortal. Follow link below
to inspect the default file and format used.</li>
<li>Clinical data should be added last for overall sample checks to
work. For example, if there is expression data for patients 1-20 and cna
data patients 15-20, it can more informatively warn about any
missing/mismatches.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">clinical_data</span> <span class="op">&lt;-</span> <span class="st">"select * from syn43278088"</span></span>
<span><span class="va">ref_map</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/nf-osi/nf-metadata-dictionary/main/mappings/cBioPortal.yaml"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/cbp_add_clinical.html">cbp_add_clinical</a></span><span class="op">(</span><span class="va">clinical_data</span>, <span class="va">ref_map</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="validation">Validation<a class="anchor" aria-label="anchor" href="#validation"></a>
</h2>
<p>There are additional steps such as generating case lists and
validation that have to be done <em>outside</em> of the package with a
cBioPortal backend, where each portal may have specific configurations
(such as genomic reference) to validate against. See the <a href="https://docs.cbioportal.org/using-the-dataset-validator/" class="external-link">general
docs for dataset validation</a>.</p>
<p>For the <em>public</em> portal, the suggested step using the public
server is given below.</p>
<p>Assuming your present working directory is
<code>~/datahub/public</code> and a study folder called
<code>npst_nfosi_ntap_2022</code> has been placed into it, mount the
dataset into the container and run validation like:</p>
<pre><code>STUDY=npst_nfosi_ntap_2022
sudo docker run --rm -v $(pwd):/datahub cbioportal/cbioportal:5.4.7 validateStudies.py -d /datahub -l $STUDY -u http://cbioportal.org -html /datahub/$STUDY/html_report</code></pre>
<p>The html report will list issues by data types to help with any
corrections needed.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Robert Allaway, Anh Nguyet Vu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
