<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • nfportalutils</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nfportalutils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9500</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/annotate-nf-processed-data.html">Annotating nextflow processed data</a></li>
    <li><a class="dropdown-item" href="../articles/bringing-portal-data-to-other-platforms-cbioportal.html">Bringing Portal Data to Other Platforms: cBioPortal</a></li>
    <li><a class="dropdown-item" href="../articles/revalidation-workflows.html">Revalidation workflows</a></li>
    <li><a class="dropdown-item" href="../articles/survey-public-files.html">Surveying public files in the portal</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/nf-osi/nfportalutils/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="table-utils">Table Utils<a class="anchor" aria-label="anchor" href="#table-utils"></a></h2>




    </div><div class="section level2">

      <h3 id="main-portal-table-data-update-and-management">Main PORTAL table data update and management<a class="anchor" aria-label="anchor" href="#main-portal-table-data-update-and-management"></a></h3>
      <div class="section-desc"><p>Augment and update one of the main portal tables (e.g. Portal - Studies)</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="add_publication_from_pubmed.html">add_publication_from_pubmed()</a></code>

        </dt>
        <dd>Add a publication to the publication table</dd>
      </dl><dl><dt>

          <code><a href="add_publication_from_unpaywall.html">add_publication_from_unpaywall()</a></code>

        </dt>
        <dd>Add a publication or preprint to the publication table via the Unpaywall API.</dd>
      </dl><dl><dt>

          <code><a href="add_publications_from_file.html">add_publications_from_file()</a></code>

        </dt>
        <dd>Add a batch of publications from spreadsheet</dd>
      </dl><dl><dt>

          <code><a href="assign_study_data_types.html">assign_study_data_types()</a></code>

        </dt>
        <dd>Summarize data types for the study</dd>
      </dl><dl><dt>

          <code><a href="calculate_related_studies.html">calculate_related_studies()</a></code>

        </dt>
        <dd>Calculate and add related studies to study table</dd>
      </dl><dl><dt>

          <code><a href="register_study.html">register_study()</a></code>

        </dt>
        <dd>Register a NEW project for the NF Data Portal in <strong>Portal - Project View</strong></dd>
      </dl><dl><dt>

          <code><a href="add_people_from_table.html">add_people_from_table()</a></code>

        </dt>
        <dd>Update the People table from a source Table or View column</dd>
      </dl><dl><dt>

          <code><a href="register_study_files.html">register_study_files()</a></code>

        </dt>
        <dd>Register a project's files in <strong>Portal - Files</strong></dd>
      </dl><dl><dt>

          <code><a href="summarize_attribute.html">summarize_attribute()</a></code>

        </dt>
        <dd>Helper summarization util</dd>
      </dl></div><div class="section level2">

      <h3 id="lower-level-table-maintenance">Lower-level table maintenance<a class="anchor" aria-label="anchor" href="#lower-level-table-maintenance"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="adjust_view.html">adjust_view()</a></code>

        </dt>
        <dd>Adjust view</dd>
      </dl><dl><dt>

          <code><a href="swap_col.html">swap_col()</a></code>

        </dt>
        <dd>Swap out old column for a new column in a schema</dd>
      </dl><dl><dt>

          <code><a href="byte_budget.html">byte_budget()</a></code>

        </dt>
        <dd>Calculate byte budget for a schema</dd>
      </dl></div><div class="section level2">
      <h2 id="project-configuration-and-management">Project Configuration and Management<a class="anchor" aria-label="anchor" href="#project-configuration-and-management"></a></h2>

      <div class="section-desc"><p>Create or retrofit an NF project to expected default structure and assets</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="new_project.html">new_project()</a></code>

        </dt>
        <dd>Create a new project</dd>
      </dl><dl><dt>

          <code><a href="add_default_fileview.html">add_default_fileview()</a></code>

        </dt>
        <dd>Create default project fileview</dd>
      </dl><dl><dt>

          <code><a href="add_default_folders.html">add_default_folders()</a></code>

        </dt>
        <dd>Create default folders</dd>
      </dl><dl><dt>

          <code><a href="make_admin.html">make_admin()</a></code>

        </dt>
        <dd>Make a user or group full admin of a Synapse entity</dd>
      </dl></div><div class="section level2">
      <h2 id="metadata-utils">Metadata Utils<a class="anchor" aria-label="anchor" href="#metadata-utils"></a></h2>




    </div><div class="section level2">

      <h3 id="general-annotations">General annotations<a class="anchor" aria-label="anchor" href="#general-annotations"></a></h3>
      <div class="section-desc"><p>Add and manage annotations on Synapse entities</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="update_study_annotations.html">update_study_annotations()</a></code>

        </dt>
        <dd>Updates a set of files with project-level annotations.</dd>
      </dl><dl><dt>

          <code><a href="annotate_with_manifest.html">annotate_with_manifest()</a></code>

        </dt>
        <dd>Set annotations from a manifest</dd>
      </dl><dl><dt>

          <code><a href="copy_annotations.html">copy_annotations()</a></code>

        </dt>
        <dd>Copy annotations</dd>
      </dl><dl><dt>

          <code><a href="dot-modify_annotation.html">.modify_annotation()</a></code>

        </dt>
        <dd>Modify a single annotation on a single file</dd>
      </dl><dl><dt>

          <code><a href="meta_qc_dataset.html">meta_qc_dataset()</a></code>

        </dt>
        <dd>QC dataset metadata with pass/fail result</dd>
      </dl><dl><dt>

          <code><a href="meta_qc_project.html">meta_qc_project()</a></code>

        </dt>
        <dd>QC metadata at the project level with pass/fail result</dd>
      </dl><dl><dt>

          <code><a href="manifest_generate.html">manifest_generate()</a></code>

        </dt>
        <dd>Generate manifest via schematic service</dd>
      </dl><dl><dt>

          <code><a href="manifest_validate.html">manifest_validate()</a></code>

        </dt>
        <dd>Validate manifest via schematic service</dd>
      </dl><dl><dt>

          <code><a href="manifest_passed.html">manifest_passed()</a></code>

        </dt>
        <dd>Provide a pass/fail summary result</dd>
      </dl><dl><dt>

          <code><a href="precheck_manifest.html">precheck_manifest()</a></code>

        </dt>
        <dd>Precheck a manifest</dd>
      </dl><dl><dt>

          <code><a href="remanifest.html">remanifest()</a></code>

        </dt>
        <dd>Reconstitute a manifest</dd>
      </dl><dl><dt>

          <code><a href="infer_data_type.html">infer_data_type()</a></code>

        </dt>
        <dd>Infer data type of a dataset folder</dd>
      </dl></div><div class="section level2">

      <h3 id="special-annotation-of-nextflow-processed-data">Special annotation of nextflow processed data<a class="anchor" aria-label="anchor" href="#special-annotation-of-nextflow-processed-data"></a></h3>
      <div class="section-desc"><p>Special annotation of nextflow processed data</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="map_reports_sarek.html">map_reports_sarek()</a></code>

        </dt>
        <dd>Map out Sarek <em>report</em> files</dd>
      </dl><dl><dt>

          <code><a href="map_sample_input_ss.html">map_sample_input_ss()</a></code>

        </dt>
        <dd>Parse nextflow samplesheet for sample inputs</dd>
      </dl><dl><dt>

          <code><a href="map_sample_io.html">map_sample_io()</a></code>

        </dt>
        <dd>Map sample input-output</dd>
      </dl><dl><dt>

          <code><a href="map_sample_output_rnaseq.html">map_sample_output_rnaseq()</a></code>

        </dt>
        <dd>Map sample to output from nf-rnaseq</dd>
      </dl><dl><dt>

          <code><a href="map_sample_output_sarek.html">map_sample_output_sarek()</a></code>

        </dt>
        <dd>Map sample to output from nf-sarek</dd>
      </dl><dl><dt>

          <code><a href="annotate_aligned_reads.html">annotate_aligned_reads()</a></code>

        </dt>
        <dd>Annotate processed aligned reads</dd>
      </dl><dl><dt>

          <code><a href="annotate_called_variants.html">annotate_called_variants()</a></code>

        </dt>
        <dd>Annotate somatic or germline variants output</dd>
      </dl><dl><dt>

          <code><a href="annotate_expression.html">annotate_expression()</a></code>

        </dt>
        <dd>Annotate processed expression output</dd>
      </dl><dl><dt>

          <code><a href="annotate_cnv.html">annotate_cnv()</a></code>

        </dt>
        <dd>Annotate CNV outputs from CNVKit</dd>
      </dl><dl><dt>

          <code><a href="annotate_reports_sarek.html">annotate_reports_sarek()</a></code>

        </dt>
        <dd>Annotate Sarek reports</dd>
      </dl><dl><dt>

          <code><a href="annotate_with_tool_stats.html">annotate_with_tool_stats()</a></code>

        </dt>
        <dd>Make annotations from workflow tool stats</dd>
      </dl><dl><dt>

          <code><a href="nf_workflow_version.html">nf_workflow_version()</a></code>

        </dt>
        <dd>Return workflow version according to workflow meta</dd>
      </dl></div><div class="section level2">
      <h2 id="dataset-creation-and-management">Dataset Creation and Management<a class="anchor" aria-label="anchor" href="#dataset-creation-and-management"></a></h2>




    </div><div class="section level2">

      <h3 id="general-dataset-creation-and-citation">General dataset creation and citation<a class="anchor" aria-label="anchor" href="#general-dataset-creation-and-citation"></a></h3>
      <div class="section-desc"><p>Create datasets in general</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="new_dataset.html">new_dataset()</a></code>

        </dt>
        <dd>Create new dataset with given items</dd>
      </dl></div><div class="section level2">

      <h3 id="specialized-dataset-creation">Specialized dataset creation<a class="anchor" aria-label="anchor" href="#specialized-dataset-creation"></a></h3>
      <div class="section-desc"><p>Create specialized datasets for nextflow processed data, i.e. with some custom construction queries and title templating.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="nf_sarek_datasets.html">nf_sarek_datasets()</a></code>

        </dt>
        <dd>Create datasets for Sarek-called somatic or germline variants results</dd>
      </dl><dl><dt>

          <code><a href="nf_star_salmon_datasets.html">nf_star_salmon_datasets()</a></code>

        </dt>
        <dd>Create dataset for STAR-Salmon expression quantification results</dd>
      </dl><dl><dt>

          <code><a href="nf_cnv_dataset.html">nf_cnv_dataset()</a></code>

        </dt>
        <dd>Create dataset for CNVKit results</dd>
      </dl></div><div class="section level2">

      <h3 id="working-with-dataset-collections-to-manage-datasets-after-creation">Working with dataset collections to manage datasets after creation<a class="anchor" aria-label="anchor" href="#working-with-dataset-collections-to-manage-datasets-after-creation"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="add_to_collection.html">add_to_collection()</a></code>

        </dt>
        <dd>Add to collection</dd>
      </dl><dl><dt>

          <code><a href="use_latest_in_collection.html">use_latest_in_collection()</a></code>

        </dt>
        <dd>Update item versions to "latest" in a collection</dd>
      </dl><dl><dt>

          <code><a href="update_items.html">update_items()</a></code>

        </dt>
        <dd>Apply updates to current collection of items</dd>
      </dl></div><div class="section level2">
      <h2 id="data-model-utils">Data Model Utils<a class="anchor" aria-label="anchor" href="#data-model-utils"></a></h2>

      <div class="section-desc"><p>Talk to a JSON-LD data model important to the portal data (i.e. NF-metadata-dictionary)</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="get_by_prop_from_json_schema.html">get_by_prop_from_json_schema()</a></code>

        </dt>
        <dd>Look up connected nodes by specified property in JSON-LD schema</dd>
      </dl><dl><dt>

          <code><a href="get_dependency_from_json_schema.html">get_dependency_from_json_schema()</a></code>

        </dt>
        <dd>Get dependencies for node in JSON-LD schema</dd>
      </dl><dl><dt>

          <code><a href="get_valid_values_from_json_schema.html">get_valid_values_from_json_schema()</a></code>

        </dt>
        <dd>Retrieve valid subclasses of a value in a JSON-LD schema</dd>
      </dl><dl><dt>

          <code><a href="key_label_to_id.html">key_label_to_id()</a></code>

        </dt>
        <dd>Query for schema key id given label</dd>
      </dl><dl><dt>

          <code><a href="schema_max_str_len.html">schema_max_str_len()</a></code>

        </dt>
        <dd>Consult schema about max string length</dd>
      </dl></div><div class="section level2">
      <h2 id="governance-utils">Governance Utils<a class="anchor" aria-label="anchor" href="#governance-utils"></a></h2>

      <div class="section-desc"><p>Analyze and manage data access/restrictions</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="make_public_viewable.html">make_public_viewable()</a></code>

        </dt>
        <dd>Set public access to VIEW (READ) only for an entity</dd>
      </dl><dl><dt>

          <code><a href="make_public.html">make_public()</a></code>

        </dt>
        <dd>Make public</dd>
      </dl><dl><dt>

          <code><a href="check_access.html">check_access()</a></code>

        </dt>
        <dd>Check access</dd>
      </dl><dl><dt>

          <code><a href="summarize_file_access.html">summarize_file_access()</a></code>

        </dt>
        <dd>Summarize file access for files within some view</dd>
      </dl><dl><dt>

          <code><a href="grant_specific_file_access.html">grant_specific_file_access()</a></code>

        </dt>
        <dd>Provide access to a specific set of files using a query result.</dd>
      </dl></div><div class="section level2">
      <h2 id="search-utils">Search Utils<a class="anchor" aria-label="anchor" href="#search-utils"></a></h2>

      <div class="section-desc"><p>Help locate Synapse accessions, etc.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="find_child.html">find_child()</a></code>

        </dt>
        <dd>Find id of a child entity in a container</dd>
      </dl><dl><dt>

          <code><a href="find_child_type.html">find_child_type()</a></code>

        </dt>
        <dd>Find children of type</dd>
      </dl><dl><dt>

          <code><a href="find_data_root.html">find_data_root()</a></code>

        </dt>
        <dd>Find data folder</dd>
      </dl><dl><dt>

          <code><a href="find_in.html">find_in()</a></code>

        </dt>
        <dd>Find in path</dd>
      </dl><dl><dt>

          <code><a href="find_nf_asset.html">find_nf_asset()</a></code>

        </dt>
        <dd>Find a standard nextflow workflow output asset</dd>
      </dl></div><div class="section level2">
      <h2 id="provenance-utils">Provenance Utils<a class="anchor" aria-label="anchor" href="#provenance-utils"></a></h2>

      <div class="section-desc"><p>Manage provenance metadata</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="add_activity.html">add_activity()</a></code>

        </dt>
        <dd>Add activity to entity</dd>
      </dl><dl><dt>

          <code><a href="add_activity_batch.html">add_activity_batch()</a></code>

        </dt>
        <dd>Add activity to multiple entities</dd>
      </dl><dl><dt>

          <code><a href="delete_provenance.html">delete_provenance()</a></code>

        </dt>
        <dd>Remove provenance info</dd>
      </dl></div><div class="section level2">
      <h2 id="content-utils">Content Utils<a class="anchor" aria-label="anchor" href="#content-utils"></a></h2>

      <div class="section-desc"><p>Create and manage content for projects and pages</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="add_default_wiki.html">add_default_wiki()</a></code>

        </dt>
        <dd>Add default wiki</dd>
      </dl><dl><dt>

          <code><a href="wiki_mod.html">wiki_mod()</a></code>

        </dt>
        <dd>Add markup to a project wiki</dd>
      </dl><dl><dt>

          <code><a href="remove_wiki_subpage.html">remove_wiki_subpage()</a></code>

        </dt>
        <dd>Remove a subpage from a project wiki</dd>
      </dl><dl><dt>

          <code><a href="data_curator_app_subpage.html">data_curator_app_subpage()</a></code>

        </dt>
        <dd>Create NF Data Curator App subpage</dd>
      </dl><dl><dt>

          <code><a href="get_project_wiki.html">get_project_wiki()</a></code>

        </dt>
        <dd>Get wiki content of synapse project(s)</dd>
      </dl><dl><dt>

          <code><a href="check_wiki_links.html">check_wiki_links()</a></code>

        </dt>
        <dd>Check wiki links</dd>
      </dl><dl><dt>

          <code><a href="remove_button.html">remove_button()</a></code>

        </dt>
        <dd>Remove button from a project wiki</dd>
      </dl></div><div class="section level2">

      <h3 id="figures-and-diagrams">Figures and diagrams<a class="anchor" aria-label="anchor" href="#figures-and-diagrams"></a></h3>
      <div class="section-desc"><p>Supplemental figures and diagrams that go into Wikis or other places</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="processing_flowchart.html">processing_flowchart()</a></code>

        </dt>
        <dd>Wrapper to create data-driven flowchart with pretty processing provenance mermaid template</dd>
      </dl><dl><dt>

          <code><a href="dsp_dataset_mapping.html">dsp_dataset_mapping()</a></code>

        </dt>
        <dd>Wrapper to create Data Sharing Plan to project dataset comparison chart</dd>
      </dl><dl><dt>

          <code><a href="bipartite_mmd_template.html">bipartite_mmd_template()</a></code>

        </dt>
        <dd>Simple bipartite representation in mermaid charts</dd>
      </dl></div><div class="section level2">
      <h2 id="export-data-to-other-platforms">Export Data to Other Platforms<a class="anchor" aria-label="anchor" href="#export-data-to-other-platforms"></a></h2>

      <div class="section-desc"><p>Helpers to export/release NF data to other platforms/databases.</p></div>


    </div><div class="section level2">

      <h3 id="cbioportal">cBioPortal<a class="anchor" aria-label="anchor" href="#cbioportal"></a></h3>
      <div class="section-desc"><p>Export data as a cBioPortal study</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="cbp_new_study.html">cbp_new_study()</a></code>

        </dt>
        <dd>Initialize a new cBioPortal study dataset</dd>
      </dl><dl><dt>

          <code><a href="cbp_new_cancer_type.html">cbp_new_cancer_type()</a></code>

        </dt>
        <dd>Create reference file for new cancer type</dd>
      </dl><dl><dt>

          <code><a href="cbp_add_maf.html">cbp_add_maf()</a></code>

        </dt>
        <dd>Export and add mutations data to cBioPortal dataset</dd>
      </dl><dl><dt>

          <code><a href="cbp_add_clinical.html">cbp_add_clinical()</a></code>

        </dt>
        <dd>Export and add clinical data to cBioPortal dataset</dd>
      </dl><dl><dt>

          <code><a href="cbp_add_expression.html">cbp_add_expression()</a></code>

        </dt>
        <dd>Export and add expression data to cBioPortal dataset</dd>
      </dl><dl><dt>

          <code><a href="cbp_add_cna.html">cbp_add_cna()</a></code>

        </dt>
        <dd>Export and add CNA (seg) data to cBioPortal dataset</dd>
      </dl></div><div class="section level2">
      <h2 id="quality-control-and-testing-utils">Quality Control and Testing Utils<a class="anchor" aria-label="anchor" href="#quality-control-and-testing-utils"></a></h2>

      <div class="section-desc"><p>QC data</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="check_readpair_validity.html">check_readpair_validity()</a></code>

        </dt>
        <dd>Check fastq read pair matches samplesheet read pair assignment.</dd>
      </dl><dl><dt>

          <code><a href="identify_read_pair.html">identify_read_pair()</a></code>

        </dt>
        <dd>Identify read pair from string</dd>
      </dl><dl><dt>

          <code><a href="test_failed.html">test_failed()</a></code>

        </dt>
        <dd>Format a test fail message.</dd>
      </dl><dl><dt>

          <code><a href="test_passed.html">test_passed()</a></code>

        </dt>
        <dd>Format a test passed message.</dd>
      </dl></div><div class="section level2">
      <h2 id="basic-utils">Basic Utils<a class="anchor" aria-label="anchor" href="#basic-utils"></a></h2>

      <div class="section-desc"><p>Low-level functions</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="syn_login.html">syn_login()</a></code>

        </dt>
        <dd>Logs into Synapse.</dd>
      </dl><dl><dt>

          <code><a href="table_query.html">table_query()</a></code>

        </dt>
        <dd>Generic table query</dd>
      </dl><dl><dt>

          <code><a href="as_table_schema.html">as_table_schema()</a></code>

        </dt>
        <dd>Transform table data to target schema for Synapse storage</dd>
      </dl><dl><dt>

          <code><a href="make_folder.html">make_folder()</a></code>

        </dt>
        <dd>Create project folders</dd>
      </dl><dl><dt>

          <code><a href="add_to_scope.html">add_to_scope()</a></code>

        </dt>
        <dd>Add to scope</dd>
      </dl><dl><dt>

          <code><a href="new_view.html">new_view()</a></code>

        </dt>
        <dd>Create a view</dd>
      </dl><dl><dt>

          <code><a href="list_project_datasets.html">list_project_datasets()</a></code>

        </dt>
        <dd>List datasets in project</dd>
      </dl><dl><dt>

          <code><a href="latest_version.html">latest_version()</a></code>

        </dt>
        <dd>Get the latest version</dd>
      </dl><dl><dt>

          <code><a href="walk.html">walk()</a></code>

        </dt>
        <dd>Walk through a directory</dd>
      </dl><dl><dt>

          <code><a href="copy.html">copy()</a></code>

        </dt>
        <dd>Create copy of entity</dd>
      </dl><dl><dt>

          <code><a href="convert_to_stringlist.html">convert_to_stringlist()</a></code>

        </dt>
        <dd>Convert a delimited string to a stringlist annotation</dd>
      </dl><dl><dt>

          <code><a href="bare_syn_id.html">bare_syn_id()</a></code>

        </dt>
        <dd>Extract synapse id from URI or other string</dd>
      </dl><dl><dt>

          <code><a href="bad_url.html">bad_url()</a></code>

        </dt>
        <dd>Helper function to check urls</dd>
      </dl><dl><dt>

          <code><a href="dot-update_table_data.html">.update_table_data()</a></code>

        </dt>
        <dd>Replace/update table contents = input data must have ROW_ID and ROW_VERSION columns to update, otherwise will append data.</dd>
      </dl><dl><dt>

          <code><a href="dot-update_view_data.html">.update_view_data()</a></code>

        </dt>
        <dd>Replace/update table contents = input data must have ROW_ID, ROW_VERSION, ETAG columns to update.</dd>
      </dl><dl><dt>

          <code><a href="from_pubmed.html">from_pubmed()</a></code>

        </dt>
        <dd>Get publication metadata from PubMed</dd>
      </dl></div><div class="section level2">
      <h2 id="internal-experimental">Internal/experimental<a class="anchor" aria-label="anchor" href="#internal-experimental"></a></h2>

      <div class="section-desc"><p>Mostly meant to be internal or experimental stuff</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="dot-delim_string_to_vector.html">.delim_string_to_vector()</a></code>

        </dt>
        <dd>Convert a delimited string to vector, utility function.</dd>
      </dl><dl><dt>

          <code><a href="dot-replace_string_column_with_stringlist_column.html">.replace_string_column_with_stringlist_column()</a></code>

        </dt>
        <dd>Replace string column with stringlist column</dd>
      </dl><dl><dt>

          <code><a href="dot-store_rows.html">.store_rows()</a></code>

        </dt>
        <dd>Adds a row to a table.</dd>
      </dl><dl><dt>

          <code><a href="missing_annotation_email.html">missing_annotation_email()</a></code>

        </dt>
        <dd>Convert a delimited string to a stringlist annotation</dd>
      </dl><dl><dt>

          <code><a href="get_doi_meta.html">get_doi_meta()</a></code>

        </dt>
        <dd>Get DOI metadata if it exists</dd>
      </dl><dl><dt>

          <code><a href="cite_dataset.html">cite_dataset()</a></code>

        </dt>
        <dd>Generate example dataset citation</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Robert Allaway, Anh Nguyet Vu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

