<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Map sample to output from nf-sarek — map_sample_output_sarek • nfportalutils</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Map sample to output from nf-sarek — map_sample_output_sarek"><meta name="description" content="See https://nf-co.re/sarek. Similar to map_sample_output_rnaseq but for Sarek outputs.
Processed outputs have been seen to have variable organization, nested first by sample or by caller as
VariantCalling/&amp;lt;SAMPLE&amp;gt;/&amp;lt;CALLER&amp;gt; or as
VariantCalling/&amp;lt;CALLER&amp;gt;/&amp;lt;SAMPLE&amp;gt;.
The hierarchy/context needed for annotation can be obtained with a fileview and given the
necessary starting point syn_out, which is the id of the relevant output folder
(called VariantCalling or variant_calling usually)."><meta property="og:description" content="See https://nf-co.re/sarek. Similar to map_sample_output_rnaseq but for Sarek outputs.
Processed outputs have been seen to have variable organization, nested first by sample or by caller as
VariantCalling/&amp;lt;SAMPLE&amp;gt;/&amp;lt;CALLER&amp;gt; or as
VariantCalling/&amp;lt;CALLER&amp;gt;/&amp;lt;SAMPLE&amp;gt;.
The hierarchy/context needed for annotation can be obtained with a fileview and given the
necessary starting point syn_out, which is the id of the relevant output folder
(called VariantCalling or variant_calling usually)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nfportalutils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9600</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/annotate-data-intro.html">Introduction to utils for annotating data</a></li>
    <li><a class="dropdown-item" href="../articles/annotate-nf-processed-data.html">Annotating nextflow processed data</a></li>
    <li><a class="dropdown-item" href="../articles/bringing-portal-data-to-other-platforms-cbioportal.html">Bringing Portal Data to Other Platforms: cBioPortal</a></li>
    <li><a class="dropdown-item" href="../articles/revalidation-workflows.html">Revalidation workflows</a></li>
    <li><a class="dropdown-item" href="../articles/survey-public-files.html">Surveying public files in the portal</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/nf-osi/nfportalutils/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Map sample to output from nf-sarek</h1>
      <small class="dont-index">Source: <a href="https://github.com/nf-osi/nfportalutils/blob/develop/R/nextflow_annotation_utils.R" class="external-link"><code>R/nextflow_annotation_utils.R</code></a></small>
      <div class="d-none name"><code>map_sample_output_sarek.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>See https://nf-co.re/sarek. Similar to <code><a href="map_sample_output_rnaseq.html">map_sample_output_rnaseq</a></code> but for Sarek outputs.
Processed outputs have been seen to have variable organization, nested first by sample or by caller as
<code>VariantCalling/&lt;SAMPLE&gt;/&lt;CALLER&gt;</code> or as
<code>VariantCalling/&lt;CALLER&gt;/&lt;SAMPLE&gt;</code>.
The hierarchy/context needed for annotation can be obtained with a fileview and given the
necessary starting point <code>syn_out</code>, which is the id of the relevant output folder
(called <code>VariantCalling</code> or <code>variant_calling</code> usually).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">map_sample_output_sarek</span><span class="op">(</span></span>
<span>  <span class="va">syn_out</span>,</span>
<span>  <span class="va">fileview</span>,</span>
<span>  output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CNVkit"</span>, <span class="st">"DeepVariant"</span>, <span class="st">"Strelka2"</span>, <span class="st">"Mutect2"</span>, <span class="st">"FreeBayes"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-syn-out">syn_out<a class="anchor" aria-label="anchor" href="#arg-syn-out"></a></dt>
<dd><p>Syn id of variant calling output folder.</p></dd>


<dt id="arg-fileview">fileview<a class="anchor" aria-label="anchor" href="#arg-fileview"></a></dt>
<dd><p>An existing fileview to use (usually the project's local fileview)
that scopes outputs and has "default" columns (<code>id</code>, <code>name</code>, <code>type</code>, <code>parentId</code>, <code>path</code>, ...). See details.</p></dd>


<dt id="arg-output">output<a class="anchor" aria-label="anchor" href="#arg-output"></a></dt>
<dd><p>Which output to select and annotate.
Defaults to checking the presence of possible prioritized outputs
("CNVkit", "DeepVariant", "Strelka2", "Mutect2", and "FreeBayes"),
though typically only a subset makes sense and can be explicitly specified, which also speeds up results.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>data.table</code> with cols <code>caller</code> <code>path</code> <code>output_name</code> <code>output_id</code> <code>sample</code> <code>workflow</code>.
An attribute <code>workflow=nf-sarek</code> will be set on the returned list,
and elements will have attribute <code>outputFrom</code> set, e.g. <code>outputFrom=CNVkit</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Note: Depending on when <code>map_sample_output_sarek</code> is run on the output directory,
there may be just <code>vcf</code> or both <code>vcf</code> and <code>maf</code> present. <strong><code>maf</code> are ignored since
technically it is not an output of sarek but from further processing with nf-vcf2maf.</strong></p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Robert Allaway, Anh Nguyet Vu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

