<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Annotating nextflow processed data • nfportalutils</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Annotating nextflow processed data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nfportalutils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9600</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/annotate-data-intro.html">Introduction to utils for annotating data</a></li>
    <li><a class="dropdown-item" href="../articles/annotate-nf-processed-data.html">Annotating nextflow processed data</a></li>
    <li><a class="dropdown-item" href="../articles/bringing-portal-data-to-other-platforms-cbioportal.html">Bringing Portal Data to Other Platforms: cBioPortal</a></li>
    <li><a class="dropdown-item" href="../articles/revalidation-workflows.html">Revalidation workflows</a></li>
    <li><a class="dropdown-item" href="../articles/survey-public-files.html">Surveying public files in the portal</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/nf-osi/nfportalutils/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Annotating nextflow processed data</h1>
            
            <h4 data-toc-skip class="date">2022-10-17</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nf-osi/nfportalutils/blob/develop/vignettes/annotate-nf-processed-data.Rmd" class="external-link"><code>vignettes/annotate-nf-processed-data.Rmd</code></a></small>
      <div class="d-none name"><code>annotate-nf-processed-data.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="intro">Intro<a class="anchor" aria-label="anchor" href="#intro"></a>
</h2>
<div class="section level3">
<h3 id="purpose">Purpose<a class="anchor" aria-label="anchor" href="#purpose"></a>
</h3>
<p>This vignette documents in-practice usage of the annotation utils for nf-processed data files.</p>
<p>The outputs are:<br>
1. A metadata manifest for a processed dataset that can be further validated with schematic before submission. 2. Provenance meta.</p>
<p>Typically, these are inspected/validated before submitting to Synapse as a final followup step.</p>
<p>Examples can be run with READ access only to processed outputs, but requires you to have DOWNLOAD access or a local copy of the input samplesheet. To actually apply any annotations of course requires EDIT access.</p>
</div>
<div class="section level3">
<h3 id="general-idea">General idea<a class="anchor" aria-label="anchor" href="#general-idea"></a>
</h3>
<p><img src="https://raw.githubusercontent.com/nf-core/rnaseq/3.14.0//docs/images/nf-core-rnaseq_metro_map_grey.png" alt="nf-core-rnaseq_metro_map" width="800"></p>
<p>A nextflow workflow generates different types of outputs along the steps in the workflow (see figure above). At some of these steps/stops, we have products that can be collected and packaged into “level 2, 3, or 4” datasets. For example, the <code>.bam/bai</code> outputs from SAMtools represent a “level 2” semi-processed dataset with a dataType of <code>AlignedReads</code>.</p>
<p>Ideally, we would simply like to point to the main folder containing all processed output files and get back a list of manifests that represent all useful dataset products from the workflow. (What is a “useful”, selectable data product is encoded in this annotation workflow.)</p>
<p>These manifests can then be used to annotate the files as well as for creation of <a href="https://help.synapse.org/docs/Datasets.2611281979.html" class="external-link">Synapse Datasets</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="set-up">Set up<a class="anchor" aria-label="anchor" href="#set-up"></a>
</h2>
<p>First load the <code>nfportalutils</code> package and log in. The recommended default usage of <code>syn_login</code> is to use it without directly passing in credentials. Instead, have available the <code>SYNAPSE_AUTH_TOKEN</code> environment variable with your token stored therein.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/nf-osi/nfportalutils" class="external-link">nfportalutils</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/syn_login.html">syn_login</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="steps">Steps<a class="anchor" aria-label="anchor" href="#steps"></a>
</h3>
<p><strong>The general annotation workflow steps are:</strong><br>
1. Parse the input samplesheet.<br>
2. Get basic context of processed outputs from the workflow run.<br>
Because none of the indexed-back output files have annotations, we have to first construct initial info <strong>sample</strong>, <strong>caller</strong>, etc. out of the good ol’ folder hierarchy and file names.<br>
3. Now with some sample and workflow context at least, link input-output data appropriately, check sample correspondence, and get into format expected for downstream.<br>
4. Transfer other meta from input to output processed files (most important are <code>individualID</code>, basic individual attributes, <code>assay</code>).<br>
5. Set annotations for processed data type based on workflow default rules.</p>
<p><strong>Some potential issues should be noted:</strong></p>
<ul>
<li>If input files have missing or incorrect annotations, processed files will have missing or incorrect annotations.<br>
</li>
<li>If sample ids and other information are updated on the original raw input files, data must be reannotated by rerunning the pipeline.<br>
</li>
<li>Anything that deviates from a relatively standard workflow run, leading to changes in locations or naming of outputs, might yield poor results or require more manual composition of steps. Standard organization and naming of files is very important.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="nf-rnaseq">nf-rnaseq<a class="anchor" aria-label="anchor" href="#nf-rnaseq"></a>
</h2>
<div class="section level3">
<h3 id="what-does-output-look-like">What does output look like?<a class="anchor" aria-label="anchor" href="#what-does-output-look-like"></a>
</h3>
<p>Use <code><a href="../reference/map_sample_output_rnaseq.html">?map_sample_output_rnaseq</a></code> to see which outputs are handled in the parameter <code>output</code>. <strong>But note that depending on how the workflow was run and data indexed back into Synapse, actual output availability may differ.</strong></p>
<p>In some projects, bam/bai files may not even be indexed back into Synapse. As an illustrative example, the workflow outputs here <strong>does not</strong> include <strong>featureCounts</strong>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">syn_out</span> <span class="op">&lt;-</span> <span class="st">"syn57382909"</span></span>
<span><span class="va">fileview</span> <span class="op">&lt;-</span> <span class="st">"syn11601495"</span></span>
<span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_sample_output_rnaseq.html">map_sample_output_rnaseq</a></span><span class="op">(</span><span class="va">syn_out</span>, <span class="va">fileview</span><span class="op">)</span> <span class="co"># check outputs only</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">o</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="what-does-input-look-like">What does input look like?<a class="anchor" aria-label="anchor" href="#what-does-input-look-like"></a>
</h3>
<p>Like output, input is just another index of files and is actually the samplesheet used by the workflow to know what files to process. Samplesheets should be public and placed in the <code>pipeline_info</code> directory as part of the workflow (most of the time).</p>
<p><strong>IMPORTANT: The samplesheet needs to be standard enough to parse correctly, i.e. to extract valid file Synapse ids from the first fastq.</strong> We use the same helper to parse samplesheets for the two workflows (both RNA-seq and Sarek), and the function will do its best to handle slight variations in samplesheet formats. Here are examples of what will work vs not:</p>
<ul>
<li>✔ OK. Excerpt from real samplesheet <a href="https://www.synapse.org/Synapse:syn51525432" class="external-link">syn51525432</a>.</li>
</ul>
<pre><code><span><span class="co">#&gt;                 sample     fastq_1     fastq_2 strandedness</span></span>
<span><span class="co">#&gt; 1 JH-2-019-DB5EH-C461C syn15261791 syn15261900         auto</span></span>
<span><span class="co">#&gt; 2 JH-2-007-B14BB-AG2A6 syn15261974 syn15262033         auto</span></span>
<span><span class="co">#&gt; 3 JH-2-009-518B9-77BH3 syn15262157 syn15262216         auto</span></span></code></pre>
<ul>
<li>✔ OK. Excerpt from real samplesheet <a href="https://www.synapse.org/Synapse:syn63172939" class="external-link">syn63172939</a>.</li>
</ul>
<pre><code><span><span class="co">#&gt;   subject sex status     sample              lane            fastq1 fastq2</span></span>
<span><span class="co">#&gt; 1  JHU002  XY      1 JHU002-043 JHU002-043-Lane-1 syn://syn22091879     NA</span></span>
<span><span class="co">#&gt; 2  JHU002  XY      1 JHU002-048 JHU002-048-Lane-1 syn://syn22091925     NA</span></span>
<span><span class="co">#&gt; 3  JHU023  XY      1 JHU023-044 JHU023-044-Lane-1 syn://syn22091973     NA</span></span>
<span><span class="co">#&gt;     datasetId   projectId output_destination_id Germline Somatic</span></span>
<span><span class="co">#&gt; 1 syn29783617 syn11638893           syn29429576        Y      NA</span></span>
<span><span class="co">#&gt; 2 syn29783617 syn11638893           syn29429576        Y      NA</span></span>
<span><span class="co">#&gt; 3 syn29783617 syn11638893           syn29429576        Y      NA</span></span></code></pre>
<ul>
<li>✖ No. Adapted from real samplesheet <a href="https://www.synapse.org/Synapse:syn63172939" class="external-link">syn63172939</a>. This will give an error because “x6” is not a valid Synapse ID. A manually corrected samplesheet will have to be provided.</li>
</ul>
<pre><code><span><span class="co">#&gt;               sample single_end</span></span>
<span><span class="co">#&gt; 1 patient10tumor1_T1          0</span></span>
<span><span class="co">#&gt; 2 patient10tumor2_T1          0</span></span>
<span><span class="co">#&gt; 3 patient10tumor3_T1          0</span></span>
<span><span class="co">#&gt;                                                             fastq_1</span></span>
<span><span class="co">#&gt; 1         s3://some-tower-bucket/syn40134517/x6/SL106309_1.fastq.gz</span></span>
<span><span class="co">#&gt; 2 s3://some-tower-bucket/syn40134517/syn7989846/SL106310_1.fastq.gz</span></span>
<span><span class="co">#&gt; 3 s3://some-tower-bucket/syn40134517/syn7989852/SL106311_1.fastq.gz</span></span>
<span><span class="co">#&gt;                                                             fastq_2</span></span>
<span><span class="co">#&gt; 1 s3://some-tower-bucket/syn40134517/syn7989839/SL106309_2.fastq.gz</span></span>
<span><span class="co">#&gt; 2 s3://some-tower-bucket/syn40134517/syn7989847/SL106310_2.fastq.gz</span></span>
<span><span class="co">#&gt; 3 s3://some-tower-bucket/syn40134517/syn7989856/SL106311_2.fastq.gz</span></span>
<span><span class="co">#&gt;   strandedness</span></span>
<span><span class="co">#&gt; 1         auto</span></span>
<span><span class="co">#&gt; 2         auto</span></span>
<span><span class="co">#&gt; 3         auto</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="connecting-input-and-output-to-automate-filled-manifests">Connecting input and output to automate filled manifests<a class="anchor" aria-label="anchor" href="#connecting-input-and-output-to-automate-filled-manifests"></a>
</h3>
<p>In contrast with the previous example, run this other example for an output directory that <em>does</em> all types of outputs we’re looking for in an nf-rnaseq workflow. This will be used for the rest of the demo. (Review the source code for <code>processed_meta</code> to see the steps encapsulated.)</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">samplesheet</span> <span class="op">&lt;-</span> <span class="st">"syn51408030"</span></span>
<span><span class="va">syn_out</span> <span class="op">&lt;-</span> <span class="st">"syn51476810"</span></span>
<span><span class="va">fileview</span> <span class="op">&lt;-</span> <span class="st">"syn11601481"</span></span>
<span><span class="va">wf_link</span> <span class="op">&lt;-</span> <span class="st">"https://nf-co.re/rnaseq/3.11.2/output#star-and-salmon"</span></span>
<span></span>
<span><span class="va">input</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_sample_input_ss.html">map_sample_input_ss</a></span><span class="op">(</span><span class="va">samplesheet</span><span class="op">)</span></span>
<span><span class="co"># Alternatively, use a local file if not on Synapse:</span></span>
<span><span class="co"># input &lt;-  map_sample_input_ss("~/work/samplesheet.csv")</span></span>
<span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_sample_output_rnaseq.html">map_sample_output_rnaseq</a></span><span class="op">(</span><span class="va">syn_out</span>, <span class="va">fileview</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">output</span><span class="op">)</span></span></code></pre></div>
<p>Generate the manifests and inspect an example result:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/processed_meta.html">processed_meta</a></span><span class="op">(</span><span class="va">input</span>, <span class="va">output</span>, workflow_link <span class="op">=</span> <span class="va">wf_link</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta</span><span class="op">$</span><span class="va">manifests</span><span class="op">$</span><span class="va">SAMtools</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="submit-manifest">Submit manifest<a class="anchor" aria-label="anchor" href="#submit-manifest"></a>
</h3>
<p>Manifests can be submitted with <a href="https://schematic.api.sagebionetworks.org/v1/ui/#" class="external-link">schematic</a>-compatible or using <code>annotate_with_manifest</code> as shown below.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mannifest_1</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">manifests</span><span class="op">$</span><span class="va">SAMtools</span></span>
<span><span class="fu"><a href="../reference/annotate_with_manifest.html">annotate_with_manifest</a></span><span class="op">(</span><span class="va">manifest_1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-provenance">Add provenance<a class="anchor" aria-label="anchor" href="#add-provenance"></a>
</h3>
<p>Provenance is basically an annotation, though treated somewhat differently in Synapse. In the result <code>meta</code> object, there is something called <code>sample_io</code> that can be provided to <code>add_activity_batch</code> to add provenance.</p>
<p>“Workflow” provides the general name to the activity, while “workflow link” provides a more persistent reference to some version/part of the workflow, which others can follow the link to get details.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">sample_io</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">sample_io</span></span>
<span><span class="va">prov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_activity_batch.html">add_activity_batch</a></span><span class="op">(</span><span class="va">sample_io</span><span class="op">$</span><span class="va">output_id</span>, </span>
<span>                           <span class="va">sample_io</span><span class="op">$</span><span class="va">workflow</span>, </span>
<span>                           <span class="va">wf_link</span>,</span>
<span>                           <span class="va">sample_io</span><span class="op">$</span><span class="va">input_id</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-dataset">Create dataset<a class="anchor" aria-label="anchor" href="#create-dataset"></a>
</h3>
<p>To create a <a href="https://help.synapse.org/docs/Datasets.2611281979.html" class="external-link">Synapse Dataset</a>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">items</span> <span class="op">&lt;-</span> <span class="va">manifest_1</span><span class="op">$</span><span class="va">entityId</span></span>
<span><span class="va">project</span> <span class="op">&lt;-</span> <span class="st">"your-dev-project-synid"</span></span>
<span><span class="va">dataset_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/new_dataset.html">new_dataset</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"STAR Salmon Gene Expression Quantification from RNA-seq"</span>,</span>
<span>                         parent <span class="op">=</span> <span class="va">project</span>,</span>
<span>                         items <span class="op">=</span> <span class="va">items</span>,</span>
<span>                         dry_run <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="nf-sarek">nf-sarek<a class="anchor" aria-label="anchor" href="#nf-sarek"></a>
</h2>
<p>The only difference is usage of <code>map_sample_output_sarek</code> at Step 2.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">samplesheet</span> <span class="op">&lt;-</span> <span class="st">"syn38793905"</span> <span class="co"># samplesheet can be stored on Synapse or locally</span></span>
<span><span class="va">syn_out</span> <span class="op">&lt;-</span> <span class="st">"syn27650634"</span></span>
<span><span class="va">fileview</span> <span class="op">&lt;-</span> <span class="st">"syn13363852"</span></span>
<span></span>
<span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_sample_input_ss.html">map_sample_input_ss</a></span><span class="op">(</span><span class="va">samplesheet</span><span class="op">)</span> <span class="co">#1</span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_sample_output_sarek.html">map_sample_output_sarek</a></span><span class="op">(</span><span class="va">syn_out</span>, <span class="va">fileview</span><span class="op">)</span> <span class="co">#2</span></span>
<span><span class="va">sarek_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/processed_meta.html">processed_meta</a></span><span class="op">(</span><span class="va">i</span>, <span class="va">o</span>, workflow_link <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View first manifest</span></span>
<span><span class="va">sarek_meta</span><span class="op">$</span><span class="va">manifests</span><span class="op">$</span><span class="va">Strelka2</span></span></code></pre></div>
<div class="section level3">
<h3 id="add-provenance-1">Add provenance<a class="anchor" aria-label="anchor" href="#add-provenance-1"></a>
</h3>
<p>Use the manifest to add provenance.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">wf_link</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FreeBayes <span class="op">=</span> <span class="st">"https://nf-co.re/sarek/3.2.3/output#freebayes"</span>,</span>
<span>             Mutect2 <span class="op">=</span> <span class="st">"https://nf-co.re/sarek/3.2.3/output#gatk-mutect2"</span>,</span>
<span>             Strelka2 <span class="op">=</span> <span class="st">"https://nf-co.re/sarek/3.2.3/output#strelka2"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/add_activity_batch.html">add_activity_batch</a></span><span class="op">(</span><span class="va">sample_io</span><span class="op">$</span><span class="va">output_id</span>, </span>
<span>                   <span class="va">sample_io</span><span class="op">$</span><span class="va">workflow</span>, </span>
<span>                   <span class="va">wf_link</span><span class="op">[</span><span class="va">sample_io</span><span class="op">$</span><span class="va">workflow</span><span class="op">]</span>, </span>
<span>                   <span class="va">sample_io</span><span class="op">$</span><span class="va">input_id</span><span class="op">)</span></span>
<span>   </span></code></pre></div>
<p>After provenance, the rest of the workflow for manifest submission or creating datasets is like the nf-rnaseq example.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Robert Allaway, Anh Nguyet Vu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
