<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="nfportalutils">
<title>Annotating nextflow processed data • nfportalutils</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Annotating nextflow processed data">
<meta property="og:description" content="nfportalutils">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">nfportalutils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9300</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/annotate-nf-processed-data.html">Annotating nextflow processed data</a>
    <a class="dropdown-item" href="../articles/bringing-portal-data-to-other-platforms-cbioportal.html">Bringing Portal Data to Other Platforms (cBioPortal)</a>
    <a class="dropdown-item" href="../articles/survey-public-files.html">Surveying public files in the portal</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nf-osi/nfportalutils/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Annotating nextflow processed data</h1>
            
            <h4 data-toc-skip class="date">2022-10-17</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nf-osi/nfportalutils/blob/HEAD/vignettes/annotate-nf-processed-data.Rmd" class="external-link"><code>vignettes/annotate-nf-processed-data.Rmd</code></a></small>
      <div class="d-none name"><code>annotate-nf-processed-data.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="intro">Intro<a class="anchor" aria-label="anchor" href="#intro"></a>
</h2>
<p>This vignette documents in-practice usage of the annotation utils for
nf-processed data files; it will probably updated as practices and
underlying functions are refined.</p>
<p>If you copy and paste some code to try out, you will need at least
read access. The code to modify the entities’ provenance/annotations are
not evaluated.</p>
</div>
<div class="section level2">
<h2 id="important-note">Important note<a class="anchor" aria-label="anchor" href="#important-note"></a>
</h2>
<p>You may notice one minor difference in the annotation utils for
RNA-seq expression vs variant calling data. The main difference is that
one will add workflow directly, while the other doesn’t. This is a
design change that will be made more consistent across the annotation
functions going forward, where workflow values will be added for
everything by default, and functions will be kept up to date to make
consistency easy for DCC staff. For non-nf-processed data, values can
still be overridden.</p>
</div>
<div class="section level2">
<h2 id="set-up">Set up<a class="anchor" aria-label="anchor" href="#set-up"></a>
</h2>
<p>First load the <code>nfportalutils</code> package and log in. The
recommended default usage of <code>syn_login</code> is to use it without
directly passing in credentials. Instead, have available the
<code>SYNAPSE_AUTH_TOKEN</code> environment variable with your token
stored therein.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/nf-osi/nfportalutils" class="external-link">nfportalutils</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/syn_login.html">syn_login</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="nf-rnaseq">nf-rnaseq<a class="anchor" aria-label="anchor" href="#nf-rnaseq"></a>
</h2>
<div class="section level3">
<h3 id="creating-helper-resource-files">Creating helper resource files<a class="anchor" aria-label="anchor" href="#creating-helper-resource-files"></a>
</h3>
<p>To add annotations and provenance, we need a reference mapping of
inputs and outputs involved in the workflow. The below puts together a
table representing this mapping by parsing the samplesheet in
combination with the output directory (“star_salmon”). For the nextflow
rna-seq workflow, the samplesheet is an output file in
“pipeline_info”.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_io</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_sample_io.html">map_sample_io</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"nf-rnaseq"</span>,</span>
<span>                           samplesheet <span class="op">=</span> <span class="st">"syn30841441"</span>, <span class="co"># synapse file</span></span>
<span>                           syn_out <span class="op">=</span> <span class="st">"syn30840584"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-provenance">Add provenance<a class="anchor" aria-label="anchor" href="#add-provenance"></a>
</h3>
<p>Add provenance for the files involved using
<code>add_activity_batch</code>.</p>
<p>Note: Why is workflow name already available but not workflow link?
The workflow link is a more specific reference and includes version
numbers.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">wf_link</span> <span class="op">&lt;-</span> <span class="st">"https://nf-co.re/rnaseq/3.7/output#star-and-salmon"</span></span>
<span><span class="va">prov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_activity_batch.html">add_activity_batch</a></span><span class="op">(</span><span class="va">sample_io</span><span class="op">$</span><span class="va">output_id</span>, </span>
<span>                           <span class="va">sample_io</span><span class="op">$</span><span class="va">workflow</span>, </span>
<span>                           <span class="va">wf_link</span>,</span>
<span>                           <span class="va">sample_io</span><span class="op">$</span><span class="va">input_id</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-annotations">Add annotations<a class="anchor" aria-label="anchor" href="#add-annotations"></a>
</h3>
<p>Annotations expected for aligned reads output are different from
quantified expression output. This is handled by a couple of specialized
functions.</p>
<div class="section level4">
<h4 id="aligned-reads-data-files">Aligned reads data files<a class="anchor" aria-label="anchor" href="#aligned-reads-data-files"></a>
</h4>
<p>First, it helps to see the template for the aligned
reads<code>(.bam</code>) data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">template</span> <span class="op">&lt;-</span> <span class="st">"bts:ProcessedAlignedReadsTemplate"</span></span>
<span><span class="va">schema</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/nf-osi/nf-metadata-dictionary/main/NF.jsonld"</span></span>
<span><span class="va">props</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_dependency_from_json_schema.html">get_dependency_from_json_schema</a></span><span class="op">(</span>id <span class="op">=</span> <span class="va">template</span><span class="op">)</span></span>
<span><span class="va">props</span></span></code></pre></div>
<p>This looks like a large number of different annotations to put
together (though most are optional). Fortunately, the utility
<code>annotate_aligned_reads</code> will try to do much of the work. It
will look for the <code>.bam</code> files and generate annotations for
those <code>.bam</code> files. The template and schema mentioned above
are the defaults so don’t have to be specified explicitly. It’s
recommended to read the docs (<code><a href="../reference/annotate_aligned_reads.html">?annotate_aligned_reads</a></code>) for
details on how annotations are compiled; usage of
<code>annotate_aligned_reads</code> does still require knowing what
should be filled in manually.</p>
<p>Note: Why is workflow info added as both provenance and annotations?
Isn’t that redundant? The answer is that the portal uses only
annotations and not provenance/annotations are more queryable
currently.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">l2_manifest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotate_aligned_reads.html">annotate_aligned_reads</a></span><span class="op">(</span><span class="va">sample_io</span>,</span>
<span>                                      samtools_stats_file <span class="op">=</span> <span class="st">"syn30841284"</span>,</span>
<span>                                      update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">l2_manifest</span><span class="op">[</span>, <span class="va">genomicReference</span> <span class="op">:=</span> <span class="st">"GRCh38"</span><span class="op">]</span></span>
<span><span class="va">l2_manifest</span><span class="op">[</span>, <span class="va">workflowLink</span> <span class="op">:=</span> <span class="st">"https://nf-co.re/rnaseq/3.7/output#star-and-salmon"</span><span class="op">]</span></span></code></pre></div>
<p>This looks good, so we can go ahead and apply the annotations:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/annotate_with_manifest.html">annotate_with_manifest</a></span><span class="op">(</span><span class="va">l2_manifest</span><span class="op">)</span></span></code></pre></div>
<p>TO DO: Instead of submitting as above, if we prefer to submit via
schematic/DCA, we can make this a schematic-compatible manifest.</p>
</div>
<div class="section level4">
<h4 id="expression-data-files">Expression data files<a class="anchor" aria-label="anchor" href="#expression-data-files"></a>
</h4>
<p>Another convenience function is provided to help annotate gene
expression data files.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">l3_manifest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotate_expression.html">annotate_expression</a></span><span class="op">(</span><span class="va">sample_io</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">l3_manifest</span><span class="op">[</span>, <span class="va">workflowLink</span> <span class="op">:=</span> <span class="st">"https://nf-co.re/rnaseq/3.7/output#star-and-salmon"</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/annotate_with_manifest.html">annotate_with_manifest</a></span><span class="op">(</span><span class="va">l3_manifest</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="nf-sarek">nf-sarek<a class="anchor" aria-label="anchor" href="#nf-sarek"></a>
</h2>
<p>As the second part of this vignette will show, the process for
variant data is not much different, though mapping inputs-outputs can be
somewhat slower.</p>
<div class="section level3">
<h3 id="creating-helper-resource-files-1">Creating helper resource files<a class="anchor" aria-label="anchor" href="#creating-helper-resource-files-1"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">sample_io</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_sample_io.html">map_sample_io</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"nf-sarek"</span>,</span>
<span>                           samplesheet <span class="op">=</span> <span class="st">"PilotBatch-WES-samplesheet.txt"</span>, <span class="co"># local file</span></span>
<span>                           syn_out <span class="op">=</span> <span class="st">"syn27650634"</span><span class="op">)</span></span>
<span><span class="co"># Modify workflow assignment a bit</span></span>
<span><span class="va">sample_io</span><span class="op">[</span><span class="va">workflow</span> <span class="op">==</span><span class="st">"Strelka"</span>, <span class="va">workflow</span> <span class="op">:=</span> <span class="st">"Strelka2"</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-provenance-1">Add provenance<a class="anchor" aria-label="anchor" href="#add-provenance-1"></a>
</h3>
<p>Use the manifest to add provenance.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">wf_link</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FreeBayes <span class="op">=</span> <span class="st">"https://nf-co.re/sarek/2.7.1/output#freebayes"</span>,</span>
<span>             Mutect2 <span class="op">=</span> <span class="st">"https://nf-co.re/sarek/2.7.1/output#gatk-mutect2"</span>,</span>
<span>             Strelka2 <span class="op">=</span> <span class="st">"https://nf-co.re/sarek/2.7.1/output#strelka2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/add_activity_batch.html">add_activity_batch</a></span><span class="op">(</span><span class="va">sample_io</span><span class="op">$</span><span class="va">output_id</span>, </span>
<span>                   <span class="va">sample_io</span><span class="op">$</span><span class="va">workflow</span>, </span>
<span>                   <span class="va">wf_link</span><span class="op">[</span><span class="va">sample_io</span><span class="op">$</span><span class="va">workflow</span><span class="op">]</span>, </span>
<span>                   <span class="va">sample_io</span><span class="op">$</span><span class="va">input_id</span><span class="op">)</span></span>
<span>   </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-annotations-1">Add annotations<a class="anchor" aria-label="anchor" href="#add-annotations-1"></a>
</h3>
<p>The <code>annotate_called_variants</code> util is used for both
somatic and germline results, and will automatically try to fill in the
right values. However, you can set these in the parameters directly –
see <code><a href="../reference/annotate_called_variants.html">?annotate_called_variants</a></code>. This returns a manifest for
examination and correction has needed.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">manifest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotate_called_variants.html">annotate_called_variants</a></span><span class="op">(</span><span class="va">sample_io</span><span class="op">)</span></span></code></pre></div>
<p>For example, if you used a slightly different version/fork, you
should update it manually in the manifest. Otherwise, apply the
annotations:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/annotate_with_manifest.html">annotate_with_manifest</a></span><span class="op">(</span><span class="va">manifest</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Robert Allaway, Anh Nguyet Vu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
